src/full/Agda/Compiler/Malfunction/Compiler.hs

```diff
@@ -0,0 +1,627 @@
+lookupConRep ns = Map.lookup (qnameNameId ns) <$> asks _conMap
+
+-- Unit is treated as a glorified value in Treeless, luckily it's fairly
+-- straight-forward to encode using the scheme described in the documentation
+-- for `translateCon`.
+unitT :: Term
+unitT = Mblock 0 []
+
+translateName :: MonadReader Env m => QName -> m Term
+translateName qn = Mvar <$> nameToIdent qn
+
+-- | Translate a Treeless name to a valid identifier in Malfunction
+--
+-- Not all names in agda are valid names in Treleess. Valid names in Agda are
+-- given by [1]. Valid identifiers in Malfunction is subject to change:
+--
+-- "Atoms: sequences of ASCII letters, digits, or symbols (the exact set of
+-- allowed symbols isn't quite nailed down yet)"[2]
+--
+-- [1. The Agda Wiki]: <http://wiki.portal.chalmers.se/agda/pmwiki.php?n=ReferenceManual2.Identifiers>
+-- [2. Malfunction Spec]: <https://github.com/stedolan/malfunction/blob/master/docs/spec.md>
+nameToIdent :: MonadReader Env m => QName -> m Ident
+nameToIdent qn = nameIdToIdent (qnameNameId qn)
+
+nameIdToIdent' :: NameId -> Maybe String -> Ident
+nameIdToIdent' (NameId a b) msuffix = (hex a ++ "." ++ hex b ++ suffix)
+  where
+    suffix = maybe "" ('.':) msuffix
+    hex = (`showHex` "") . toInteger
+
+nameIdToIdent :: MonadReader Env m => NameId -> m Ident
+nameIdToIdent nid = do
+  x <- Map.lookup nid <$> asks _qnameConcreteMap
+  return (nameIdToIdent' nid x)
+
+-- | Translates a treeless identifier to a malfunction identifier.
+qnameNameId :: QName -> NameId
+qnameNameId = nameId . qnameName
+
+-- | Compiles treeless "bindings" to a malfunction module given groups of defintions.
+compile
+  :: Env                -- ^ Environment.
+  -> [(QName, TTerm)] -- ^ List of treeless bindings.
+  -> Mod
+compile env bs = runTranslate (compileM bs) env
+
+
+compileM :: MonadReader Env m => [(QName, TTerm)] -> m Mod
+compileM allDefs = do
+  bs <- mapM translateSCC recGrps
+  return $ MMod bs []
+  where
+    translateSCC scc = case scc of
+      AcyclicSCC single -> uncurry translateBinding single
+      CyclicSCC grp -> translateMutualGroup grp
+    recGrps :: [SCC (QName, TTerm)]
+    recGrps = dependencyGraph allDefs
+
+translateMutualGroup :: MonadReader Env m => [(QName, TTerm)] -> m Binding
+translateMutualGroup bs = Recursive <$> mapM (uncurry translateBindingPair) bs
+
+translateBinding :: MonadReader Env m => QName -> TTerm -> m Binding
+translateBinding q t = uncurry Named <$> translateBindingPair q t
+
+translateBindingPair :: MonadReader Env m => QName -> TTerm -> m (Ident, Term)
+translateBindingPair q t = do
+  iden <- nameToIdent q
+  (\t' -> (iden, t')) <$> translateTerm t
+
+dependencyGraph :: [(QName, TTerm)] -> [SCC (QName, TTerm)]
+dependencyGraph qs = stronglyConnComp [ ((qn, tt), qnameNameId qn, edgesFrom tt)
+                                    | (qn, tt) <- qs ]
+  where edgesFrom = Set.toList . qnamesIdsInTerm
+
+
+qnamesIdsInTerm :: TTerm -> Set NameId
+qnamesIdsInTerm t = go t mempty
```

Comment:
I think this can be implemented on top of the the ``NamesIn`` class. I don't think there is an instance for ``TTerm`` already, but that could be added.

https://github.com/agda/agda/blob/c297fe095a76ddcd00766df6eeffdf958d272273/src/full/Agda/Syntax/Internal/Names.hs#L26